### PR TITLE
Modified ModelTask::findBelongsTo() to check for table referenced by id field

### PR DESCRIPTION
**Pull request by djamesfar**
Fix for issue #6004

### CODE REVIEW

File: src/Shell/Task/ModelTask.php [hunk 1]
```diff
@@ -253,6 +254,12 @@ public function findBelongsTo($model, array $associations)
                 ];
             } else {
                 $tmpModelName = $this->_modelNameFromKey($fieldName);
+
+                if (!in_array(Inflector::tableize($tmpModelName), $this->_tables)){
+                    $found = $this->findTableReferencedBy($fieldName);
+                    if ($found) $tmpModelName = Inflector::camelize($found);
```
Comment (markstory) [inline src/Shell/Task/ModelTask.php]: Control structures need braces.

File: src/Shell/Task/ModelTask.php [hunk 2]
```diff
@@ -268,6 +275,34 @@ public function findBelongsTo($model, array $associations)
...
+     */
+    public function findTableReferencedBy($keyField = null){
+        $db = ConnectionManager::get($this->connection);
+        $schema = $db->schemaCollection();
+        $tables = $schema->listTables();                                // get tables in db
+
+        foreach ($tables as $table){
+            $meta = $schema->describe($table, ['forceRefresh'=>true]);  // get table meta data
+            $columns = $meta->columns();
+            if (!in_array($keyField, $columns)) continue;               // not in this table
```
Comment (markstory) [inline src/Shell/Task/ModelTask.php]: More braces please.

File: src/Shell/Task/ModelTask.php [hunk 3]
```diff
@@ -268,6 +275,34 @@ public function findBelongsTo($model, array $associations)
...
+            $constraints = $meta->constraints();                        // found it; get constraints
+            foreach ($constraints as $constraint){
+                $constraint_info = $meta->constraint($constraint);      // get details of constraint
+                if(in_array($keyField, $constraint_info['columns'])) {  // match?
```
Comment (markstory) [inline src/Shell/Task/ModelTask.php]: Inline comments go one the line above. Most of these comments are not necessary though as the code is pretty straightforward.

**Comment by djamesfar** [comment]
so remove all comments?

File: src/Shell/Task/ModelTask.php [hunk 4]
```diff
@@ -268,6 +277,36 @@ public function findBelongsTo($model, array $associations)
...
 
     /**
+     * find the table, if any, actually referenced by the passed key field.
+     *   Search tables in db for keyField; if found search key constraints
+     *   for the table to which it refers.
+     *
+     * @param null $keyField
+     * @return null
+     */
+    public function findTableReferencedBy($keyField = null){
```
Comment (dereuromark) [inline src/Shell/Task/ModelTask.php]: missing space before {

File: src/Shell/Task/ModelTask.php [hunk 5]
```diff
@@ -268,6 +277,36 @@ public function findBelongsTo($model, array $associations)
...
 
     /**
+     * find the table, if any, actually referenced by the passed key field.
+     *   Search tables in db for keyField; if found search key constraints
+     *   for the table to which it refers.
+     *
+     * @param null $keyField
+     * @return null
+     */
+    public function findTableReferencedBy($keyField = null) {
```
Comment (markstory) [inline src/Shell/Task/ModelTask.php]: Braces need to be on the following like for functions.  Eg.

```php
public function findTableReferencedBy($keyField = null)
{
```

File: src/Shell/Task/ModelTask.php [hunk 6]
```diff
@@ -268,6 +277,36 @@ public function findBelongsTo($model, array $associations)
...
+        foreach ($tables as $table){
+            $meta = $schema->describe($table, ['forceRefresh'=>true]);
+            $columns = $meta->columns();
+            if (!in_array($keyField, $columns)){
+                continue;
+            }
+            $constraints = $meta->constraints();
+            foreach ($constraints as $constraint){
+                $constraint_info = $meta->constraint($constraint);
+                if(in_array($keyField, $constraint_info['columns'])) {
```
Comment (markstory) [inline src/Shell/Task/ModelTask.php]: Should be `if (`.

File: src/Shell/Task/ModelTask.php [hunk 7]
```diff
@@ -268,6 +277,36 @@ public function findBelongsTo($model, array $associations)
...
+     */
+    public function findTableReferencedBy($keyField = null) {
+        $db = ConnectionManager::get($this->connection);
+        $schema = $db->schemaCollection();
+        $tables = $schema->listTables();
+
+        foreach ($tables as $table){
+            $meta = $schema->describe($table, ['forceRefresh'=>true]);
+            $columns = $meta->columns();
+            if (!in_array($keyField, $columns)){
```
Comment (markstory) [inline src/Shell/Task/ModelTask.php]: Should be `)) {`.

File: src/Shell/Task/ModelTask.php [hunk 8]
```diff
@@ -268,6 +277,36 @@ public function findBelongsTo($model, array $associations)
...
+     *
+     * @param null $keyField
+     * @return null
+     */
+    public function findTableReferencedBy($keyField = null) {
+        $db = ConnectionManager::get($this->connection);
+        $schema = $db->schemaCollection();
+        $tables = $schema->listTables();
+
+        foreach ($tables as $table){
```
Comment (markstory) [inline src/Shell/Task/ModelTask.php]: Should be `) {`.

File: src/Shell/Task/ModelTask.php [hunk 9]
```diff
@@ -268,6 +277,36 @@ public function findBelongsTo($model, array $associations)
...
+            $meta = $schema->describe($table, ['forceRefresh'=>true]);
+            $columns = $meta->columns();
+            if (!in_array($keyField, $columns)){
+                continue;
+            }
+            $constraints = $meta->constraints();
+            foreach ($constraints as $constraint){
+                $constraint_info = $meta->constraint($constraint);
```
Comment (markstory) [inline src/Shell/Task/ModelTask.php]: Variable names can't have `_` in them.

**Comment by markstory** [comment]
The [PHPCS build](https://travis-ci.org/cakephp/bake/jobs/53475334) has a list of all the coding standard related issues.

**Comment by AD7six** [comment]
There are relevant test fails.

**Comment by djamesfar** [comment]
Although this fixes the case where bake created fake associations for _id fields that *do* have actual references to a table, but didn't or couldn't follow the naming conventions, in cases where there is a field that just happens to end in '_id' but is not a foreign key (references nothing), I cannot imagine a need for bake to create a belongsTo reference to a fake class that doesn't exist, either as a class or a table.  Thoughts?

**Comment by lorenzo** [comment]
@djamesfar if it ends in _id, there is a big chance it is referencing something. Maybe we could generate the association but comment it out in the generated code?

**Comment by djamesfar** [comment]
@lorenzo - that's an idea.  It would produce a stub that the user could uncomment and modify if they needed it.  
I think we should merge these changes for now and continue discussing that.

**Comment by lorenzo** [comment]
@djamesfar would you be able to add tests for this?

**Comment by djamesfar** [comment]
for Travis?

**Comment by lorenzo** [comment]
@djamesfar they will run in travis, yes.

**Comment by djamesfar** [comment]
I'll have to figure out how.

**Comment by markstory** [comment]
You can run the tests locally as well:

```
cd bake
php composer.phar install
vendor/bin/phpunit
```

Should run the tests against SQLite.

**Comment by djamesfar** [comment]
Thanks.  Test resulted in:

PHPUnit 4.5.0 by Sebastian Bergmann and contributors.

Configuration read from
/home/djfar/PhpstormProjects/cake3test/vendor/cakephp/bake/phpunit.xml.dist

...............................................................  63 / 210 (
30%)
............................................................... 126 / 210 (
60%)
............................................................... 189 / 210 (
90%)
.....................

Time: 3.77 seconds, Memory: 29.75Mb

OK (210 tests, 340 assertions)

How do I set this up on GitHub so it runs on travis?

**Comment by lorenzo** [comment]
It is already running on travis. Above the coments input textarea you will see a "All is well" green checkmark. If you click on "details" it will take you to the results in travis

**Comment by djamesfar** [comment]
@lorenzo - I thought you asked me to add tests for this...

**Comment by AD7six** [comment]
@djamesfar he did test _for the code added_. The pre-existing tests being run don't test code you've added.

**Comment by djamesfar** [comment]
Okay, thanks.  How do I add tests for my added code?

**Comment by lorenzo** [comment]
@djamesfar Now that we are a bit more relaxed after launching 3.0, would you like some guidance for adding tests? Otherwise we can look into doing it since I think this is a great improvement

File: src/Shell/Task/ModelTask.php [hunk 10]
```diff
@@ -268,6 +277,39 @@ public function findBelongsTo($model, array $associations)
     }
 
     /**
+     * find the table, if any, actually referenced by the passed key field.
+     *   Search tables in db for keyField; if found search key constraints
+     *   for the table to which it refers.
+     *
+     * @param null $keyField  field to look for
+     * @return null
```
Comment (dereuromark) [inline src/Shell/Task/ModelTask.php]: return type is `something|null` :)

**Comment by djamesfar** [comment]
@lorenzo Congrats on 3.0 release, it's a great product.  I will accept your offer as I have become very busy at work.  Let me know if you need help with a use case.

**Comment by markstory** [comment]
Where could I find the code for these changes? Right now github is showing 'unknown repository'. Without the repository I can finish of the tests for this.

**Comment by markstory** [comment]
Nevermind, I can re-implement the code based on the diff in github.

**Comment by djamesfar** [comment]
github.com/djamesfar/bake

**Comment by markstory** [comment]
Closing as a new pull request with tests is open.

**Comment by djamesfar** [comment]
Thanks, Mark.

**Pull request closed without merging**